cmd/govim/quickfix.go

```diff
@@ -75,20 +76,46 @@ func (v *vimstate) updateQuickfixWithDiagnostics(force bool, wasPrevNotDiagnosti
 	if !wasPrevNotDiagnostics && len(v.lastQuickFixDiagnostics) > 0 {
 		var want qflistWant
 		v.Parse(v.ChannelExpr(`getqflist({"idx":0})`), &want)
-		if want.Idx == 0 {
+		if want.Idx == 1 {
```

Comment:
@myitcv I'm not sure if it was a bug to check 0 instead of 1, but I need that check to have consistant test results.

If we try to guess the new index when `want.Idx == 1`, we have flaky results because the qf list is populated multiple times, starting with sub-parts of the final list.

Reply:
The key is a comment a few lines above (which is easy to miss, I accept)

```
// Note: indexes are 1-based, hence 0 means "no index"
```

Hence, the check here is saying "if there is nothing selected, let's leave it that way"

Reply:
Indeed, the value is 0 when the quickfix is empty, but as soon as there's at least one entry, the command returns a number >0.

But since before you check `len(v.lastQuickFixDiagnostics) > 0`, I'm pretty sure the command will never return 0 here.

I'll try to explain why skipping selection when `want.Idx == 1` matters, using as an example the test `quickfix_fallback_directory.txt`. This tests uses 4 files, all of them has an error :

- `alice/alice.go`
- `bob.go`
- `charly.go`
- `dave.go`

For unknown reason (maybe a bug), the method `updateQuickfixDiagnostics` is invoked 2 times when a file is opened in the test (`vim ex 'e charly.go'`). The first call doesn't have the full list of errors (it's random, can be any error of the 4), while the second call has the full list of errors.

Example:
- first call : the qf list is populated with only 
```json
[{"filename": "bob.go"}]
```
- second call: the qf list is populated with the full list 
```json
[{"filename": "alice/alice.go"}, {"filename": "bob.go"}, {"filename": "charly.go"}, {"filename": "dave.go"}]
```
(I skipped the other fields for clarity)

If we don't skip index selection when currentIndex == 1 (`want.Idx == 1`, we probably need to rename the variable), the second call will try to select the index using `bob.go` as a reference, and so will select index 2. This is why without this condition, the test fails from the start at line 10, when we assert that opening vim with a file with errors the default qf index is 1.

If you want to experience it yourself, just replace `want.Idx==1` by `want.Idx==0`, and run the test `quickfix_fallback_directory` multiple times, it will randomly fail according to what is the content of the sublist of errors from the first call. It will succeed only if by any chance the subpart contains the same first item as the full list (`alice/alice.go`).

Reply:
> But since before you check len(v.lastQuickFixDiagnostics) > 0, I'm pretty sure the command will never return 0 here.

I'd tend to agree. Please can you add a comment to this effect? So that if we are wrong, we can fix the code _and_ the comment?

> I'll try to explain why skipping selection when want.Idx == 1 matters

Thanks. I think I followed your explanation; but please correct me if anything that follows misses the points you were making.

> For unknown reason (maybe a bug), the method updateQuickfixDiagnostics is invoked 2 times when a file is opened in the test (vim ex 'e charly.go'). The first call doesn't have the full list of errors (it's random, can be any error of the 4), while the second call has the full list of errors.

To my mind this could just as easily happen in "normal" operation, where I'm making changes that result in an error earlier in the first (or in a file lexicographically before the file position that is the first error). i.e. there isn't anything surprising/unique about the scenario to it being a "bug".

> If we don't skip index selection when currentIndex == 1 (want.Idx == 1, we probably need to rename the variable), the second call will try to select the index using bob.go as a reference, and so will select index 2.

I think what you're saying though assumes the user hasn't somehow changed the selection in the quickfix window. Because if they have, and they intentionally chose the diagnostic at index 1 (i.e. the first one), then _not_ retaining that selection when something new comes along would be "surprising". 

Again, this comes back to a point you made earlier:

> Indeed, the value is 0 when the quickfix is empty, but as soon as there's at least one entry, the command returns a number >0.

We can't distinguish between the user having intentionally chosen the first entry from what happens automatically when there is anything in the quickfix list. Hence I think we need to retain the currently selected item. It's at least the simplest thing to explain to the user.

What are your thoughts?

Reply:
Ok thanks for the clarification about what I suspected to be a bug. I agree with you, having the index at 1 can also be intentional from the user. 

I reverted the change but I  had also to remove the first assertion on the qf index at the beginning of the test to prevent failure. I added comments explaining what we talked about.
See b3330e7d197ce8022b76b97a90ea69b83dc045f7